### PR TITLE
BitcoinJ support

### PR DESCRIPTION
**Pull request by hlandau**
This commit adds support for BitcoinJ SPV client installation to the
installer. The client must be launched manually. @JeremyRand

### CODE REVIEW

File: ncdns.nsi [hunk 1]
```diff
@@ -415,6 +480,64 @@ done:
...
+
+  # Configure ncdns to use BitcoinJ port.
+  FileOpen $4 "$INSTDIR\etc\ncdns.conf" a
+  FileSeek $4 0 END
+  FileWrite $4 '$\r$\n$\r$\n## ++SPV++$\r$\n## Added automatically by installer to point ncdns to SPV client.$\r$\nnamecoinrpcaddress="127.0.0.1:6563"$\r$\n## ++/SPV++$\r$\n$\r$\n'
+  FileClose $4
+
+  # Write a batch script to enable BitcoinJ to be easily launched.
+  FileOpen $4 "$INSTDIR\BitcoinJ\Launch BitcoinJ.cmd" w
+  FileWrite $4 'C:$\r$\ncd "C:\ProgramData\NamecoinBitcoinJ"$\r$\njava -jar "$INSTDIR\BitcoinJ\bitcoinj-daemon.jar" --connection.proxyenabled=false --connection.streamisolation=false --server.port=6563$\r$\n'
```
Comment (JeremyRand) [inline ncdns.nsi]: I'd suggest using `leveldbtxcache` mode when launching bitcoinj-daemon.  (The documentation has info on how to use that mode.)  `leveldbtxcache` takes about 10 minutes to sync, so it should be sufficiently lightweight for the foreseeable future, and it's quite a bit more secure than the headers-only modes.

Comment (hlandau) [reply ncdns.nsi]: Fixed. @JeremyRand

**Comment by JeremyRand** [comment]
Looks like the artifact download isn't yet downloading `bitcoinj-daemon.jar`?

**Comment by JeremyRand** [comment]
> Looks like the artifact download isn't yet downloading `bitcoinj-daemon.jar`?

Oh... I guess that might be because the official download is hosted by Google Drive?

**Comment by hlandau** [comment]
Heh. I didn't even know there was an official download. Currently, you need to provide your own `bitcoinj-daemon.jar`, yes.

File: ncdns.nsi [hunk 2]
```diff
@@ -173,10 +214,27 @@ Function ComponentDialogCreate
...
     ${NSD_SetText} $hCtl_components_dialog_NamecoinCore_No "I will configure Namecoin Core myself (manual configuration required)"
+
+    # Use Namecoin Core by default if it's already installed.
+    ${NSD_SetState} $hCtl_components_dialog_NamecoinCore_Yes ${BST_CHECKED}
   ${Else}
     ${NSD_SetText} $hCtl_components_dialog_NamecoinCore_Status "An existing Namecoin Core installation was not detected."
-    ${NSD_SetText} $hCtl_components_dialog_NamecoinCore_Yes "Install and configure Namecoin Core (recommended)"
+    ${If} $JREDetected == 1
+      ${NSD_SetText} $hCtl_components_dialog_NamecoinCore_Yes "Install and configure Namecoin Core"
+      ${NSD_SetText} $hCtl_components_dialog_NamecoinCore_SPV "Install and use the BitcoinJ SPV client (recommended)"
```
Comment (JeremyRand) [inline ncdns.nsi]: I'd prefer not to officially recommend the SPV client over Namecoin Core.  For now, how about labeling Namecoin Core "(heavier, more secure)" and labeling the SPV client "(lighter, less secure)"?  (At some point we might want to elaborate on what the tradeoffs are, but I think that can wait.)

File: ncdns.nsi [hunk 3]
```diff
@@ -415,6 +480,64 @@ done:
...
+  FileWrite $4 '$\r$\n$\r$\n## ++SPV++$\r$\n## Added automatically by installer to point ncdns to SPV client.$\r$\nnamecoinrpcaddress="127.0.0.1:6563"$\r$\n## ++/SPV++$\r$\n$\r$\n'
+  FileClose $4
+
+  # Write a batch script to enable BitcoinJ to be easily launched.
+  FileOpen $4 "$INSTDIR\BitcoinJ\Launch BitcoinJ.cmd" w
+  FileWrite $4 'C:$\r$\ncd "C:\ProgramData\NamecoinBitcoinJ"$\r$\njava -jar "$INSTDIR\BitcoinJ\bitcoinj-daemon.jar" --connection.proxyenabled=false --connection.streamisolation=false --namelookup.latest.algo=leveldbtxcache --server.port=6563$\r$\n'
+  FileClose $4
+
+  # Create shortcuts to the batch script.
+  CreateShortcut "$SMPROGRAMS\BitcoinJ.lnk" "$INSTDIR\BitcoinJ\Launch BitcoinJ.cmd"
```
Comment (JeremyRand) [inline ncdns.nsi]: Can we rename `BitcoinJ.lnk` to `Namecoin BitcoinJ.lnk`?

File: ncdns.nsi [hunk 4]
```diff
@@ -415,6 +480,64 @@ done:
...
+  CreateShortcut "$DESKTOP\BitcoinJ.lnk" "$INSTDIR\BitcoinJ\Launch BitcoinJ.cmd"
```
Comment (JeremyRand) [inline ncdns.nsi]: Can we rename `BitcoinJ.lnk` to `Namecoin BitcoinJ.lnk`?

**Comment by JeremyRand** [comment]
In 0006b2d2a75a0a39754dd43cdc709c4bd4b4da7d, when an existing Namecoin Core node was detected, both "Automatically configure Namecoin Core (recommended)" and "Install and use the BitcoinJ SPV client instead" are ticked by default; I assume only the former should be.

**Comment by JeremyRand** [comment]
No other changes requested for 0006b2d2a75a0a39754dd43cdc709c4bd4b4da7d.

**Comment by hlandau** [comment]
Updated.

File: ncdns.nsi (hunk 2 above)
Comment (JeremyRand) [reply ncdns.nsi]: Confirmed fixed in fe095d49f7ac0e229fdc6c3a3b6ead896d7c2a50

File: ncdns.nsi (hunk 3 above)
Comment (JeremyRand) [reply ncdns.nsi]: Confirmed fixed in fe095d49f7ac0e229fdc6c3a3b6ead896d7c2a50

File: ncdns.nsi (hunk 4 above)
Comment (JeremyRand) [reply ncdns.nsi]: Confirmed fixed in fe095d49f7ac0e229fdc6c3a3b6ead896d7c2a50

**Comment by JeremyRand** [comment]
> In 0006b2d, when an existing Namecoin Core node was detected, both "Automatically configure Namecoin Core (recommended)" and "Install and use the BitcoinJ SPV client instead" are ticked by default; I assume only the former should be.

Confirmed fixed in fe095d49f7ac0e229fdc6c3a3b6ead896d7c2a50.

**Comment by JeremyRand** [comment]
ACK fe095d49f7ac0e229fdc6c3a3b6ead896d7c2a50

**Pull request merged**